File: tests/fixtures/sample_project/billing/stripe_client.py
"""Stripe integration — charges, refunds, webhook signature verification."""
from __future__ import annotations

import hashlib
import hmac
import json
import time

from ..users.crud import get_user
from ..db.connection import get_connection
from .invoices import create_invoice


STRIPE_SECRET_KEY = "sk_test_REPLACE"
STRIPE_WEBHOOK_SECRET = "whsec_REPLACE"
WEBHOOK_TOLERANCE_SEC = 300


def charge_customer(user_id: int, amount_cents: int, description: str) -> dict:
    """Create a Stripe charge and mirror it into our billing table.

    Flow: validate user → call Stripe API → on success, create an invoice
    and write a billing record locally.
    """
    user = get_user(user_id)
    if user is None or not user.is_active:
        raise BillingError(f"user {user_id} not chargeable")

    # In a real impl this would call stripe.Charge.create(...)
    stripe_response = _fake_stripe_charge(amount_cents, description)
    charge_id = stripe_response["id"]

    conn = get_connection()
    conn.execute(
        "INSERT INTO charges (user_id, stripe_id, amount_cents, status, created_at) "
        "VALUES (?, ?, ?, ?, ?)",
        (user_id, charge_id, amount_cents, stripe_response["status"], int(time.time())),
    )
    invoice = create_invoice(user_id, amount_cents, charge_id, description)
    return {"charge_id": charge_id, "invoice_id": invoice["id"]}


def refund_charge(charge_id: str, reason: str) -> dict:
    """Issue a full refund for a previous charge and log it."""
    conn = get_connection()
    row = conn.execute(
        "SELECT user_id, amount_cents, status FROM charges WHERE stripe_id = ?",
        (charge_id,),
    ).fetchone()
    if row is None:
        raise BillingError(f"unknown charge {charge_id}")
    if row[2] == "refunded":
        raise BillingError(f"charge {charge_id} already refunded")

    # In real impl: stripe.Refund.create(charge=charge_id, reason=reason)
    refund_id = f"re_{charge_id[3:]}"

    conn.execute(
        "UPDATE charges SET status = 'refunded' WHERE stripe_id = ?",
        (charge_id,),
    )
    conn.execute(
        "INSERT INTO refunds (charge_id, refund_id, reason, created_at) VALUES (?, ?, ?, ?)",
        (charge_id, refund_id, reason, int(time.time())),
    )
    return {"refund_id": refund_id, "amount_cents": row[1]}


def verify_webhook(payload: bytes, signature_header: str) -> dict:
    """Validate a Stripe webhook signature and return the event body.

    Stripe-Signature header format: 't=timestamp,v1=hex_signature'.
    Raises WebhookVerificationError on any mismatch.
    """
    parts = dict(p.split("=", 1) for p in signature_header.split(","))
    timestamp = int(parts["t"])
    signature = parts["v1"]

    if abs(time.time() - timestamp) > WEBHOOK_TOLERANCE_SEC:
        raise WebhookVerificationError("timestamp outside tolerance")

    signed_payload = f"{timestamp}.".encode() + payload
    expected = hmac.new(
        STRIPE_WEBHOOK_SECRET.encode(), signed_payload, hashlib.sha256
    ).hexdigest()

    if not hmac.compare_digest(expected, signature):
        raise WebhookVerificationError("signature mismatch")

    return json.loads(payload)


def handle_webhook_event(event: dict) -> None:
    """Dispatch a verified Stripe webhook event to the right handler."""
    event_type = event.get("type")
    if event_type == "charge.succeeded":
        _on_charge_succeeded(event["data"]["object"])
    elif event_type == "charge.failed":
        _on_charge_failed(event["data"]["object"])
    elif event_type == "charge.refunded":
        _on_charge_refunded(event["data"]["object"])


def _on_charge_succeeded(charge: dict) -> None:
    conn = get_connection()
    conn.execute(
        "UPDATE charges SET status = 'succeeded' WHERE stripe_id = ?",
        (charge["id"],),
    )


def _on_charge_failed(charge: dict) -> None:
    conn = get_connection()
    conn.execute(
        "UPDATE charges SET status = 'failed' WHERE stripe_id = ?",
        (charge["id"],),
    )


def _on_charge_refunded(charge: dict) -> None:
    conn = get_connection()
    conn.execute(
        "UPDATE charges SET status = 'refunded' WHERE stripe_id = ?",
        (charge["id"],),
    )


def _fake_stripe_charge(amount_cents: int, description: str) -> dict:
    """Stand-in for the real Stripe API call used in the fixture."""
    return {"id": f"ch_{int(time.time())}", "amount": amount_cents, "status": "succeeded"}


class BillingError(Exception):
    """Raised when a charge or refund cannot proceed."""


class WebhookVerificationError(Exception):
    """Raised when a Stripe webhook signature does not match."""

File: tests/fixtures/sample_project_ts/src/billing/stripeClient.ts
// Stripe integration — charges, refunds, webhook signature verification.

import { createHmac, timingSafeEqual } from "crypto";

import { getConnection } from "../db/connection";
import { getUser } from "../users/crud";
import { createInvoice } from "./invoices";

const STRIPE_WEBHOOK_SECRET = "whsec_REPLACE";
const WEBHOOK_TOLERANCE_SEC = 300;

/**
 * Create a Stripe charge and mirror it into our billing table.
 *
 * Flow: validate user → call Stripe API → on success, create an invoice
 * and write a billing record locally.
 */
export function chargeCustomer(
  userId: number,
  amountCents: number,
  description: string,
): Record<string, unknown> {
  const user = getUser(userId);
  if (user === null || !user.isActive) {
    throw new BillingError(`user ${userId} not chargeable`);
  }

  const stripeResponse = fakeStripeCharge(amountCents, description);
  const chargeId = stripeResponse.id;

  const conn = getConnection();
  conn.execute(
    "INSERT INTO charges (user_id, stripe_id, amount_cents, status, created_at) VALUES (?, ?, ?, ?, ?)",
    [userId, chargeId, amountCents, stripeResponse.status, Math.floor(Date.now() / 1000)],
  );
  const invoice = createInvoice(userId, amountCents, chargeId, description);
  return { chargeId, invoiceId: invoice.id };
}

/** Issue a full refund for a previous charge and log it. */
export function refundCharge(chargeId: string, reason: string): Record<string, unknown> {
  const conn = getConnection();
  const { rows } = conn.execute(
    "SELECT user_id, amount_cents, status FROM charges WHERE stripe_id = ?",
    [chargeId],
  );
  if (rows.length === 0) {
    throw new BillingError(`unknown charge ${chargeId}`);
  }
  if (rows[0][2] === "refunded") {
    throw new BillingError(`charge ${chargeId} already refunded`);
  }
  const refundId = `re_${chargeId.slice(3)}`;
  conn.execute("UPDATE charges SET status = 'refunded' WHERE stripe_id = ?", [chargeId]);
  void reason;
  return { refundId, amountCents: rows[0][1] };
}

/**
 * Validate a Stripe webhook signature and return the event body.
 *
 * Stripe-Signature header format: 't=timestamp,v1=hex_signature'.
 * Throws WebhookVerificationError on any mismatch.
 */
export function verifyWebhook(payload: string, signatureHeader: string): Record<string, unknown> {
  const parts = Object.fromEntries(signatureHeader.split(",").map((p) => p.split("=", 2)));
  const timestamp = Number(parts.t);
  const signature = parts.v1;
  // A malformed header (missing t/v1) must surface as WebhookVerificationError,
  // not a downstream TypeError from Number(undefined)/Buffer.from(undefined).
  if (!signature || Number.isNaN(timestamp)) {
    throw new WebhookVerificationError("malformed signature header");
  }

  if (Math.abs(Date.now() / 1000 - timestamp) > WEBHOOK_TOLERANCE_SEC) {
    throw new WebhookVerificationError("timestamp outside tolerance");
  }

  const signedPayload = `${timestamp}.${payload}`;
  const expected = Buffer.from(
    createHmac("sha256", STRIPE_WEBHOOK_SECRET).update(signedPayload).digest("hex"),
  );
  const provided = Buffer.from(signature);
  // timingSafeEqual throws on unequal-length buffers; length-check first so a
  // bad signature is reported as a mismatch rather than throwing a TypeError.
  if (expected.length !== provided.length || !timingSafeEqual(expected, provided)) {
    throw new WebhookVerificationError("signature mismatch");
  }
  return JSON.parse(payload);
}

/** Dispatch a verified Stripe webhook event to the right handler. */
export function handleWebhookEvent(event: Record<string, unknown>): void {
  const eventType = event.type as string;
  const data = event.data as { object: { id: string } };
  if (eventType === "charge.succeeded") {
    onChargeStatus(data.object.id, "succeeded");
  } else if (eventType === "charge.failed") {
    onChargeStatus(data.object.id, "failed");
  } else if (eventType === "charge.refunded") {
    onChargeStatus(data.object.id, "refunded");
  }
}

function onChargeStatus(stripeId: string, status: string): void {
  const conn = getConnection();
  conn.execute("UPDATE charges SET status = ? WHERE stripe_id = ?", [status, stripeId]);
}

/** Stand-in for the real Stripe API call used in the fixture. */
function fakeStripeCharge(amountCents: number, description: string): { id: string; status: string } {
  void [amountCents, description];
  return { id: `ch_${Math.floor(Date.now() / 1000)}`, status: "succeeded" };
}

/** Raised when a charge or refund cannot proceed. */
export class BillingError extends Error {}

/** Raised when a Stripe webhook signature does not match. */
export class WebhookVerificationError extends Error {}
